src/common/sorting/hashed_sort.cpp

```diff
@@ -0,0 +1,482 @@
+		if (!unsorted) {
+			unsorted = make_uniq<ColumnDataCollection>(context.client, gstate.payload_types);
+			unsorted->InitializeAppend(unsorted_append);
+		}
+		unsorted->Append(unsorted_append, input_chunk);
+		return;
+	}
+
+	//	Payload prefix is the input data
+	payload_chunk.Reset();
+	for (column_t i = 0; i < input_chunk.ColumnCount(); ++i) {
+		payload_chunk.data[i].Reference(input_chunk.data[i]);
+	}
+
+	//	Compute any sort columns that are not references and append them to the end of the payload
+	if (!gstate.sort_exprs.empty()) {
+		sort_chunk.Reset();
+		sort_exec.Execute(input_chunk, sort_chunk);
+		for (column_t i = 0; i < sort_chunk.ColumnCount(); ++i) {
+			payload_chunk.data[input_chunk.ColumnCount() + i].Reference(sort_chunk.data[i]);
+		}
+	}
+	payload_chunk.SetCardinality(input_chunk);
+
+	//	OVER(ORDER BY...)
+	if (sort_local) {
+		auto &hash_group = *gstate.hash_groups[0];
+		OperatorSinkInput input {*hash_group.sort_global, *sort_local, interrupt};
+		hash_group.sort->Sink(context, payload_chunk, input);
+		return;
+	}
+
+	// OVER(PARTITION BY...)
+	auto &hash_vector = payload_chunk.data.back();
+	Hash(input_chunk, hash_vector);
+	for (idx_t col_idx = 0; col_idx < input_chunk.ColumnCount(); ++col_idx) {
+		payload_chunk.data[col_idx].Reference(input_chunk.data[col_idx]);
+	}
+
+	gstate.UpdateLocalPartition(local_grouping, grouping_append);
+	local_grouping->Append(*grouping_append, payload_chunk);
+}
+
+void HashedSortLocalSinkState::Combine() {
+	// Window::Combine:
+	// Sort::Sink then Sort::Combine (per hash partition)
+	// Sort::Combine
+	// ColumnDataCollection::Combine
+
+	// OVER()
+	if (sort_col_count == 0) {
+		// Only one partition again, so need a global lock.
+		lock_guard<mutex> glock(gstate.lock);
+		if (gstate.unsorted) {
+			if (unsorted) {
+				gstate.unsorted->Combine(*unsorted);
+				unsorted.reset();
+			}
+		} else {
+			gstate.unsorted = std::move(unsorted);
+		}
+		return;
+	}
+
+	//	OVER(ORDER BY...)
+	if (sort_local) {
+		auto &hash_group = *gstate.hash_groups[0];
+		auto &sort = *hash_group.sort;
+		OperatorSinkCombineInput input {*hash_group.sort_global, *sort_local, interrupt};
+		sort.Combine(context, input);
+		sort_local.reset();
+		return;
+	}
+
+	// OVER(PARTITION BY...)
+	if (!local_grouping) {
+		return;
+	}
+
+	// Flush our data and lock the bit count
+	gstate.CombineLocalPartition(local_grouping, grouping_append);
+
+	//	Don't scan the hash column
+	vector<column_t> column_ids;
+	for (column_t i = 0; i < gstate.payload_types.size(); ++i) {
+		column_ids.emplace_back(i);
+	}
+
+	//	Loop over the partitions and add them to each hash group's global sort state
+	TupleDataScanState scan_state;
+	DataChunk chunk;
+	auto &partitions = local_grouping->GetPartitions();
+	for (hash_t hash_bin = 0; hash_bin < partitions.size(); ++hash_bin) {
+		auto &partition = *partitions[hash_bin];
+		if (!partition.Count()) {
+			continue;
+		}
+
+		partition.InitializeScan(scan_state, column_ids);
```

Comment:
You can initialize this scan with `TupleDataPinProperties::DESTROY_AFTER_DONE` to destroy the partitioned data as you sink it into the sort.